### PR TITLE
[wip] Multifile support

### PR DESCRIPTION
**Pull request by victorbalan**

### CODE REVIEW

File: app/elements/coder/code-editor/code-editor.html [hunk 1]
```diff
@@ -25,7 +25,7 @@
 		<context-menu id="contextMenu"></context-menu>
 		<div class="container">
 			<div class="explorer-container" style$="{{explorerStyle}}">
-				<explorer-item id="rootFolder" node="{{tree.root}}" path=""></explorer-item>
+				<explorer-element style="height: 100%;"></explorer-element>
```
Comment (lorenzleutgeb) [inline app/elements/coder/code-editor/code-editor.html]: No inline style.

File: app/elements/coder/code-editor/file-explorer/explorer-element.html [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+					value: ''
+				}
+			},
+			listeners: {
+				'rename': 'onRename',
+				'add-file': 'onAddFile',
+				'add-folder': 'onAddFolder'
+			},
+			ready: function(){
+				this.tree = {'root': {'a': 'content', 'c': {'file': 'content', 'd': {'e': { 'f': {'g': 'content'}}}}}};
```
Comment (lorenzleutgeb) [inline app/elements/coder/code-editor/file-explorer/explorer-element.html]: Do not pre-populate the tree.

File: app/elements/coder/code-editor/file-explorer/explorer-element.html [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+				this.refreshTree(tree);
+			},
+			addItem: function(tree, path, item){
+				var path = this.splitPath(path);
+
+				var node = this.getNodeForPathArray(tree, path.slice(0, path.length - 1));
+				var name = path.pop();
+				if(!!node[name]){
+					console.log('already exists');
```
Comment (lorenzleutgeb) [inline app/elements/coder/code-editor/file-explorer/explorer-element.html]: Make this a toast or smth.

File: app/elements/coder/code-editor/file-explorer/explorer-element.html [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+			},
+			getFiles: function(){
+				var tree = this.tree;
+
+				var elements = [];
+				this.getPathsRecursively(tree, elements, '');
+				return elements.map(function(it) {
+					var blob = new Blob([it.value]);
+					blob.name = it.path;
+					console.log(blob);
```
Comment (lorenzleutgeb) [inline app/elements/coder/code-editor/file-explorer/explorer-element.html]: Delete?

File: app/elements/coder/code-editor/file-explorer/file-item.html [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+				return path + item.name;
+			},
+			startRenaming: function(){
+				// ez hack for root folder
+				if(this.item.name === 'root'){
+					return;
+				}
+				this.renaming = true;
+			},
+			_handleRename: function(e) {
```
Comment (lorenzleutgeb) [inline app/elements/coder/code-editor/file-explorer/file-item.html]: Duplicated code `_handleDialog`.

File: app/elements/coder/challenge/tasks/endpoints/canvas-game-task.html [hunk 6]
```diff
@@ -2,12 +2,8 @@
 <link rel="import" href="../../../../../../bower_components/paper-button/paper-button.html">
 <link rel="import" href="../../../../../../bower_components/paper-material/paper-material.html">
 <link rel="import" href="../../../../../../bower_components/paper-toast/paper-toast.html">
-<link rel="import" href="/elements/coder/code-editor/code-editor.html">
-<link rel="import" href="/elements/coder/code-editor/language-selector.html">
-<link rel="import" href="/elements/common/services/template-service.html">
-<link rel="import" href="/elements/tasks/task-display.html">
 <link rel="import" href="../../../../../../bower_components/iron-flex-layout/classes/iron-flex-layout.html">
-
+<!--TODO(victorbalan) adapt this task to our current model-->
```
Comment (lorenzleutgeb) [inline app/elements/coder/challenge/tasks/endpoints/canvas-game-task.html:6]: So is it deprecated now or will we maintain it?

Comment (victorbalan) [inline app/elements/coder/challenge/tasks/endpoints/canvas-game-task.html:6]: i don`t know..this is a veeeery old commit from paul. It`s the robot game.

Comment (lorenzleutgeb) [inline app/elements/coder/challenge/tasks/endpoints/canvas-game-task.html:6]: well then let's leave it in deprecated state?

File: app/elements/coder/challenge/tasks/endpoints/general-code-task.html [hunk 7]
```diff
@@ -117,15 +109,17 @@
...
+				},
+				language: {
+					type: Object,
+					notify: true
 				}
 			},
 			runCode: function(){
-				this.$.codeEditor.run();
-				this.control.loading();
+				//todo(vbalan): I am deprecated. Please fix me
```
Comment (lorenzleutgeb) [inline app/elements/coder/challenge/tasks/endpoints/general-code-task.html:119]: Deprecated code will not be fixed. It just fades out.

File: app/elements/coder/challenge/tasks/endpoints/canvas-game-task.html (hunk 6 above)
Comment (victorbalan) [inline app/elements/coder/challenge/tasks/endpoints/canvas-game-task.html:6]: yes..at least for now. I will deal with all the deprecated stuff sometime after these weeks.

File: app/elements/coder/challenge/tasks/endpoints/general-code-task.html (hunk 7 above)
Comment (victorbalan) [inline app/elements/coder/challenge/tasks/endpoints/general-code-task.html:119]: all these comments are from @pbochis

File: app/elements/coder/code-editor/code-editor.html [hunk 8]
```diff
@@ -23,94 +44,163 @@
...
 				this.$.editor.setTheme('ace/theme/chrome');
 				this.$.editor.setOptions({
 					enableBasicAutocompletion: true
 				});
-				var self = this;
+				var that = this;
+				this.$.editor.getSession().on('change', function() {
+					var content = that.$.editor.getValue();
+					if(!that.selectedFile){
+						return console.log('no selected file')
```
Comment (lorenzleutgeb) [inline app/elements/coder/code-editor/code-editor.html:86]: Maybe a toast?

Comment (victorbalan) [inline app/elements/coder/code-editor/code-editor.html:86]: this does not regard the user.

File: app/elements/coder/code-editor/file-explorer/file-explorer.html [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+				}
+			},
+			listeners: {
+				'rename': 'onRename',
+				'delete': 'onDelete',
+				'add-file': 'onAddFile',
+				'add-folder': 'onAddFolder'
+			},
+			ready: function(){
+				tree = {'a': 'content', 'c': {'file': 'content', 'd': {'e': { 'f': {'g': 'content'}}}}};
```
Comment (lorenzleutgeb) [inline app/elements/coder/code-editor/file-explorer/file-explorer.html]: You are pre-populating?

Comment (victorbalan) [inline app/elements/coder/code-editor/file-explorer/file-explorer.html]: i`ll push the fix soon :D

File: app/elements/coder/code-editor/file-explorer/file-explorer.html [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+			is: 'file-explorer',
+			properties: {
+				tree: {
+					type: Object,
+					notify: true
+				},
+				names: {
+					value: {
+						good: /^[a-zA-Z0-9-_~\.]{1,255}$/,
+						bad: ['.', '..']
```
Comment (lorenzleutgeb) [inline app/elements/coder/code-editor/file-explorer/file-explorer.html:27]: You duplicated this from `tree.js`?! Is that really necessary?

Comment (victorbalan) [inline app/elements/coder/code-editor/file-explorer/file-explorer.html:27]: oh. i will remove `tree.js` in my next commit

**Pull request merged**